File: 20-actors/matsurigoto/methods/modules/test_credential_issue.py
#!/usr/bin/env python3
"""Conformance tests for the credential-issue module (matsurigoto 政, ADR-2606052300).

The check-digit tests reproduce the published ICAO Doc 9303 worked examples exactly:
  doc number L898902C3 → 6 ; DOB 740812 → 2 ; expiry 120415 → 9
and the full UTOPIA/ERIKSSON specimen line-2. Standalone + pytest.
"""
from __future__ import annotations

import pathlib
import sys

HERE = pathlib.Path(__file__).resolve().parent
sys.path.insert(0, str(HERE))

import credential_issue as P  # noqa: E402

# canonical ICAO 9303 specimen line 2
SPECIMEN_L2 = "L898902C36UTO7408122F1204159ZE184226B<<<<<10"


def test_no_server_authority_document_unsigned():
    """G1 — SOD/proof unsigned; the issuing state signs with its ICAO-PKD key."""
    assert P.SERVER_HELD_AUTHORITY is False
    p = P.issue_passport("L898902C3", "UTO", "UTO", "ERIKSSON", "ANNA MARIA",
                         "740812", "F", "120415", "did:web:x", "ZE184226B")
    assert p["document"]["sod"] is None
    assert p["document"]["proof"] is None


def test_icao_doc_number_check_digit():
    """ICAO 9303 example: L898902C3 → 6."""
    assert P.mrz_check_digit("L898902C3") == "6"


def test_icao_dob_check_digit():
    """ICAO 9303 example: 740812 → 2."""
    assert P.mrz_check_digit("740812") == "2"


def test_icao_expiry_check_digit():
    """ICAO 9303 example: 120415 → 9."""
    assert P.mrz_check_digit("120415") == "9"


def test_filler_value_zero():
    assert P.mrz_check_digit("<<<<<<") == "0"


def test_full_specimen_line2_reproduced():
    """The complete UTOPIA / ERIKSSON ANNA MARIA TD3 line 2 is reproduced exactly."""
    mrz = P.build_td3_mrz("L898902C3", "UTO", "UTO", "ERIKSSON", "ANNA MARIA",
                          "740812", "F", "120415", "ZE184226B")
    assert mrz["line2"] == SPECIMEN_L2
    assert len(mrz["line2"]) == 44


def test_specimen_line1():
    mrz = P.build_td3_mrz("L898902C3", "UTO", "UTO", "ERIKSSON", "ANNA MARIA",
                          "740812", "F", "120415", "ZE184226B")
    assert mrz["line1"] == "P<UTOERIKSSON<<ANNA<MARIA" + "<" * (44 - len("P<UTOERIKSSON<<ANNA<MARIA"))
    assert len(mrz["line1"]) == 44


def test_validate_specimen_passes():
    assert P.validate_td3_line2(SPECIMEN_L2) is True


def test_validate_detects_corruption():
    bad = list(SPECIMEN_L2)
    bad[0] = "X" if bad[0] != "X" else "Y"
    assert P.validate_td3_line2("".join(bad)) is False


def test_validate_rejects_wrong_length():
    assert P.validate_td3_line2("TOO SHORT") is False


def test_roundtrip_arbitrary_passport_validates():
    p = P.issue_passport("AB1234567", "JPN", "JPN", "YAMADA", "TARO",
                         "900101", "M", "300101", "did:web:etz")
    assert P.validate_td3_line2(p["mrz"]["line2"]) is True


def test_bad_country_code_raises():
    try:
        P.build_td3_mrz("X", "JP", "JPN", "A", "B", "900101", "M", "300101")
    except ValueError:
        return
    raise AssertionError("2-letter country code must raise")


def test_solve_is_gated_at_r0():
    try:
        P.solve()
    except RuntimeError:
        return
    raise AssertionError("solve() must raise at R0")


def _run():
    fns = [v for k, v in sorted(globals().items()) if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
        print(f"  ok  {fn.__name__}")
    print(f"\n{len(fns)}/{len(fns)} passed")
    return 0


if __name__ == "__main__":
    sys.exit(_run())

File: 20-actors/matsurigoto/methods/modules/test_credential_issue.clj
;; test_credential_issue.clj — matsurigoto credential-issue: ICAO 9303 TD3 MRZ + 7-3-1 check-digit
;; parity with credential_issue.py. Run via `bb test:matsurigoto`. ADR-2606142300.
(ns matsurigoto.methods.modules.test-credential-issue
  (:require [clojure.test :refer [deftest is testing run-tests]]
            [matsurigoto.methods.modules.credential-issue :as ci]))

(def ^:private eriksson
  {:doc-number "L898902C3" :issuing-state "UTO" :nationality "UTO" :surname "ERIKSSON"
   :given-names "ANNA MARIA" :dob-yymmdd "740812" :sex "F" :expiry-yymmdd "120415"
   :subject-did "did:web:example" :personal-number "ZE184226B"})

(deftest icao-worked-example-check-digits
  (testing "ICAO Doc 9303 7-3-1 check digits (canonical worked example)"
    (is (= "6" (ci/mrz-check-digit "L898902C3")))   ; ICAO 9303 worked example
    (is (= "2" (ci/mrz-check-digit "740812")))       ; DOB
    (is (= "9" (ci/mrz-check-digit "120415")))))     ; expiry

(deftest eriksson-specimen-byte-identical
  (testing "ERIKSSON TD3 specimen reproduced byte-for-byte (golden from credential_issue.py)"
    (let [mrz (:mrz (ci/issue-passport eriksson))]
      (is (= "P<UTOERIKSSON<<ANNA<MARIA<<<<<<<<<<<<<<<<<<<" (:line1 mrz)))
      (is (= "L898902C36UTO7408122F1204159ZE184226B<<<<<10" (:line2 mrz)))
      (is (= 44 (count (:line1 mrz))))
      (is (= 44 (count (:line2 mrz))))
      (is (= {:doc "6" :dob "2" :expiry "9" :personal "1" :composite "0"} (:check-digits mrz)))
      (is (true? (ci/validate-td3-line2 (:line2 mrz)))))))

(deftest validation-detects-corruption
  (testing "a flipped MRZ line-2 digit fails the check"
    (let [l2 "L898902C36UTO7408122F1204159ZE184226B<<<<<10"]
      (is (true?  (ci/validate-td3-line2 l2)))
      (is (false? (ci/validate-td3-line2 (str (subs l2 0 9) "7" (subs l2 10)))))   ; doc check flipped
      (is (false? (ci/validate-td3-line2 (subs l2 0 43))))                          ; wrong length
      (is (false? (ci/validate-td3-line2 "नहीं"))))))                                 ; non-MRZ chars

(deftest unsigned-document-discipline
  (testing "G1 — document SOD/proof unsigned; G6 — only MRZ fields"
    (let [doc (:document (ci/issue-passport eriksson))]
      (is (nil? (:sod doc)))                                 ; G1 — issuing state signs the SOD
      (is (nil? (:proof doc)))                               ; G1
      (is (false? (:server-held-authority doc)))             ; G1
      (is (= "issued-unsigned" (:status doc)))
      (is (= ["VerifiableCredential" "Passport"] (:type doc))))))

(deftest build-and-issue-guards
  (testing "MRZ field guards + G1 live-issuance gate"
    (is (thrown? Exception (ci/build-td3-mrz (assoc eriksson :issuing-state "JP"))))   ; not 3-letter
    (is (thrown? Exception (ci/build-td3-mrz (assoc eriksson :dob-yymmdd "7408"))))    ; not YYMMDD
    (is (thrown? Exception (ci/build-td3-mrz (assoc eriksson :sex "X"))))              ; bad sex
    (is (thrown? Exception (ci/issue-passport (assoc eriksson :doc-number ""))))       ; no doc number
    (is (thrown? Exception (ci/issue-passport (assoc eriksson :surname ""))))          ; no surname
    (is (thrown? Exception (ci/solve)))                                                ; live issuance gated
    (is (false? ci/server-held-authority))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'matsurigoto.methods.modules.test-credential-issue)]
    (System/exit (if (pos? (+ fail error)) 1 0))))
